llvm/lib/LTO/ThinLTOCodeGenerator.cpp

```diff
@@ -1237,6 +1237,9 @@ std::string
 ThinLTOCodeGenerator::writeGeneratedObject(StringRef OutputPath,
                                            ModuleCacheEntry *CacheEntry,
                                            const MemoryBuffer &OutputBuffer) {
+  if (sys::fs::exists(OutputPath))
+    sys::fs::remove(OutputPath);
```

Comment:
Would it not be better for future maintenance to move `sys::fs::remove` at the `writeObject` function that does the hard-linking, instead of requiring that all the `writeObject` callers do it before calling the function?

Reply:
`writeObject` is overwritable by sub-class and all sub class use the parent version as fallback if their fast path doesn't work. That will make multiple `exists` check on the same output file. Just try to save `status` call to FS.

Reply:
The idea is `writeGeneratedObject` should be in charge of clearing the output path before calling the fast write object methods. The other check added in this commit is probably not necessary (just to keep parity) since it should work with or without removing destination path

Reply:
But you don't need to call `sys::fs::exists` & `sys::fs::remove` in `ModuleCacheEntry::writeObject`, that function is using `raw_fd_ostream` and can create a `raw_fd_ostream` with `sys::fs::CD_CreateAlways` option, which will replace the file if it already exists.

It seems you only need to remove it for `FileModuleCacheEntry::writeObject` which is calling `sys::fs::create_hard_link`, right?

And for `FileModuleCacheEntry::writeObject` you could consider avoid always calling `sys::fs::exists` and check if `sys::fs::create_hard_link` returned `llvm::errc::file_exists`, before removing the file.

Reply:
That is the ideal case except I am now afraid  of file system :)